README.md
<h1 align="center">each-promise
  <a href="https://www.npmjs.com/package/each-promise"><img src="https://img.shields.io/npm/v/each-promise.svg" alt="npm version"></a>
  <a href="https://www.npmjs.com/package/each-promise"><img src="https://img.shields.io/npm/l/each-promise.svg" alt="npm version"></a>
  <a href="https://www.npmjs.com/package/each-promise"><img src="https://img.shields.io/npm/dm/each-promise.svg" alt="npm downloads monthly"></a>
  <a href="https://www.npmjs.com/package/each-promise"><img src="https://img.shields.io/npm/dt/each-promise.svg" alt="npm downloads total"></a>
  <br>
<img src="https://cdn.jsdelivr.net/emojione/assets/svg/1f300.svg" width="256" height="256" alt="Each Promise - Async control flow library"><br>
Asynchronous control flow library
</h1>

> Iterate over promises, promise-returning or async/await functions in series or parallel. Support settle (fail-fast), concurrency (limiting) and hooks system (start, beforeEach, afterEach, finish)

[![codeclimate][codeclimate-img]][codeclimate-url] 
[![codestyle][standard-img]][standard-url] 
[![linux build][travis-img]][travis-url] 
[![windows build][appveyor-img]][appveyor-url] 
[![codecov][coverage-img]][coverage-url] 
[![dependency status][david-img]][david-url]

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [Background](#background)
  * [Why not "promise fun"?](#why-not-promise-fun)
  * [Why not separate libs?](#why-not-separate-libs)
- [API](#api)
  * [.serial](#serial)
  * [.parallel](#parallel)
  * [.each](#each)
- [Options](#options)
- [Hooks](#hooks)
- [Item](#item)
- [Finish hook](#finish-hook)
- [Related](#related)
- [Contributing](#contributing)
- [Building docs](#building-docs)
- [Running tests](#running-tests)
- [Author](#author)
- [Logo](#logo)
- [License](#license)

_(TOC generated by [verb](https://github.com/verbose/verb) using [markdown-toc](https://github.com/jonschlinkert/markdown-toc))_

## Install
Install with [npm](https://www.npmjs.com/)

```
$ npm install each-promise --save
```

or install using [yarn](https://yarnpkg.com)

```
$ yarn add each-promise
```

## Usage
> For more use-cases see the [tests](./test.js)

```js
const eachPromise = require('each-promise')
const arr = [
  123,
  'foo',
  () => 456,
  Promise.resolve(567)
  false,
  () => Promise.resolve(11)
]

eachPromise
  .serial(arr)
  .then((res) => {
    console.log(res) // => [123, 'foo', 456, 567, false, 11]
  })
```

## Background
You may think why this exists, what is this for, why not Sindre's microlibs like [p-map][], [p-map-series][], [p-settle][], [p-each-series][] or [p-reduce][].

### Why not "promise fun"?
They do their jobs okey, but in some cases they don't. And that's the my case. I need control over _"fast fail"_ behavior, also known as _"settle"_ or _"bail"_. I need serial and parallel iteration, but parallel with concurrency too. They requires node v4, and uses native Promise constructor. I believe in that we should not use modern things if we don't need them, it is just syntax sugar. This package is written in way that works in node versions below v4 and also you can pass custom Promise constructor through [options.Promise](#options) if you want.

- node@4 required
- no hooks system
- no settle / fail-fast / bail
- no custom Promise
- no real and meaningful tests
- concurrency control

**[back to top](#readme)**

### Why not separate libs?
Why not separate `.serial` and `.parallel` into own libs like Sindre did? Because the main core logic and difference is absolutely in just 2-3 lines of code and one `if` check. The main thing is that `parallel` uses `for` loop with `concurrency` combination, and `series` does not use loops, but recursive function calls.

For free you get hooks system. And really it cost nothing. It just able to be done, because the structure of the code and because I need such thing.

- node v0.10 and above
- custom Promise constructor
- real settle / fail fast
- hook system, through options
- very stable and well tested with real tests
- concurrency control

**[back to top](#readme)**

## API

### [.serial](index.js#L61)
> Iterate over `iterable` in series (serially) with optional `opts` (see [options section](#options)) and optional `mapper` function (see [item section](#item)).

**Params**

* `<iterable>` **{Array}**: iterable object like array with any type of values    
* `[mapper]` **{Function}**: function to apply to each item in `iterable`, see [item section](#item)    
* `[opts]` **{Object}**: see [options section](#options)    
* `returns` **{Promise}**: Always resolved or rejected promise  

**Example**

```js
var delay = require('delay')
var eachPromise = require('each-promise')

var arr = [
  () => delay(500).then(() => 1),
  () => delay(200).then(() => { throw Error('foo') }),
  () => delay(10).then(() => 3),
  () => delay(350).then(() => 4),
  () => delay(150).then(() => 5)
]

eachPromise
  .serial(arr)
  .then((res) => {
    console.log(res) // [1, Error: foo, 3, 4, 5]
  })

// see what happens when parallel
eachPromise
  .parallel(arr)
  .then((res) => {
    console.log(res) // => [3, 5, Error: foo, 4, 1]
  })

// pass `settle: false` if you want
// to stop after first error
eachPromise
  .serial(arr, { settle: false })
  .catch((err) => console.log(err)) // => Error: foo
```

### [.parallel](index.js#L145)
> Iterate concurrently over `iterable` in parallel (support limiting with `opts.concurrency`) with optional `opts` (see [options section](#options)) and optional `mapper` function (see [item section](#item)).

**Params**

* `<iterable>` **{Array}**: iterable object like array with any type of values    
* `[mapper]` **{Function}**: function to apply to each item in `iterable`, see [item section](#item)    
* `[opts]` **{Object}**: see [options section](#options)    
* `returns` **{Promise}**: Always resolved or rejected promise  

**Example**

```js
var eachPromise = require('each-promise')

var arr = [
  function one () {
    return delay(200).then(() => {
      return 123
    })
  },
  Promise.resolve('foobar'),
  function two () {
    return delay(1500).then(() => {
      return 345
    })
  },
  delay(10).then(() => 'zero'),
  function three () {
    return delay(400).then(() => {
      coffffnsole.log(3) // eslint-disable-line no-undef
      return 567
    })
  },
  'abc',
  function four () {
    return delay(250).then(() => {
      return 789
    })
  },
  function five () {
    return delay(100).then(() => {
      sasasa // eslint-disable-line no-undef
      return 444
    })
  },
  function six () {
    return delay(80).then(() => {
      return 'last'
    })
  }
]

// does not stop after first error
// pass `settle: false` if you want
eachPromise
  .parallel(arr)
  .then((res) => {
    console.log(res)
    // => [
    //   'foobar',
    //   'abc',
    //   'zero',
    //   'last',
    //   ReferenceError: sasasa is not defined,
    //   123,
    //   789,
    //   ReferenceError: coffffnsole is not defined
    //   345
    // ]
  })
```

### [.each](index.js#L193)
> Iterate over `iterable` in series or parallel (default), depending on default `opts`. Pass `opts.serial: true` if you want to iterate in series, pass `opts.serial: false` or does not pass anything for parallel.

**Params**

* `<iterable>` **{Array}**: iterable object like array with any type of values    
* `[mapper]` **{Function}**: function to apply to each item in `iterable`, see [item section](#item)    
* `[opts]` **{Object}**: see [options section](#options)    
* `returns` **{Promise}**: Always resolved or rejected promise  

**Example**

```js
var delay = require('delay')
var eachPromise = require('each-promise')

var arr = [
  123,
  function () {
    return delay(500).then(() => 456)
  },
  Promise.resolve(678),
  function () {
    return 999
  },
  function () {
    return delay(200).then(() => 'foo')
  }
]

eachPromise
  .each(arr)
  .then(function (res) {
    console.log('done', res) // => [123, 678, 999, 'foo', 456]
  })
```

## Options
> You have control over everything, through options.

* `Promise` **{Function}**: custom Promise constructor to be used, defaults to native
* `mapper` **{Function}**: function to apply to each item in `iterable`, see [item section](#item)
* `settle` **{Boolean}**: if `false` stops after first error (also known as _"fail-fast"_ or _"bail"_), default `true`
* `flat` **{Boolean}**: result array to contain only values, default `true`
* `concurrency` **{Number}**: works only with `.parallel` method, defaults to `iterable` length
* `start` **{Function}**: on start hook, see [hooks section](#hooks)
* `beforeEach` **{Function}**: called before each item in `iterable`, see [hooks section](#hooks)
* `afterEach` **{Function}**: called after each item in `iterable`, see [hooks section](#hooks)
* `finish` **{Function}**: called at the end of iteration, see [hooks section](#hooks)
* `context` **{Object}**: custom context to be passed to each `fn` in `iterable`
* `args` **{Array}**: custom argument(s) to be pass to `fn`, given value is arrayified

**[back to top](#readme)**

## Hooks
> You can do what you want between stages through hooks - start, before each, after each, finish.

* `start` **{Function}**: called at the start of iteration, before anything
* `beforeEach` **{Function}**: passed with `item, index, arr` arguments
  + `item` is an object with `value`, `reason` and `index` properties, see [item section](#item)
  + `index` is the same as `item.index`
  + `arr` is the iterable object - array or object
* `afterEach` **{Function}**: passed with `item, index, arr` arguments
  + `item` is an object with `value`, `reason` and `index` properties, see [item section](#item)
  + `index` is the same as `item.index`
  + `arr` is the iterable object - array or object
* `finish` **{Function}**: called at the end of iteration, see [finish hook section](#finish-hook)

**[back to top](#readme)**

## Item
> That object is special object, that is passed to `beforeEach` and `afterEach` hooks, also can be found in `result` object if you pass `opts.flat: false` option. And passed to `opts.mapper` function too.

* `item.value` resolved/rejected promise value, if at `beforeEach` hook it can be `function`
* `item.reason` may not exist if `item.value`, if exist it is standard Error object
* `item.index` is number, order of "executing", not the order that is defined in `iterable`

**[back to top](#readme)**

## Finish hook
> This hooks is called when everything is finished / completed. At the very end of iteration. It is passed with `err, result` arguments where:

* `err` is an Error object, if `opts.settle: false`, otherwise `null`
* `result` is always an array with values or [item objects](#item) if `opts.flat: false`

**[back to top](#readme)**

## Related
- [always-done](https://www.npmjs.com/package/always-done): Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A… [more](https://github.com/hybridables/always-done#readme) | [homepage](https://github.com/hybridables/always-done#readme "Handle completion and errors with elegance! Support for streams, callbacks, promises, child processes, async/await and sync functions. A drop-in replacement for [async-done][] - pass 100% of its tests plus more")
- [minibase-create-plugin](https://www.npmjs.com/package/minibase-create-plugin): Utility for [minibase][] and [base][] that helps you create plugins | [homepage](https://github.com/node-minibase/minibase-create-plugin#readme "Utility for [minibase][] and [base][] that helps you create plugins")
- [minibase-is-registered](https://www.npmjs.com/package/minibase-is-registered): Plugin for [minibase][] and [base][], that adds `isRegistered` method to your application to detect if plugin is already… [more](https://github.com/node-minibase/minibase-is-registered#readme) | [homepage](https://github.com/node-minibase/minibase-is-registered#readme "Plugin for [minibase][] and [base][], that adds `isRegistered` method to your application to detect if plugin is already registered and returns true or false if named plugin is already registered on the instance.")
- [minibase](https://www.npmjs.com/package/minibase): Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the… [more](https://github.com/node-minibase/minibase#readme) | [homepage](https://github.com/node-minibase/minibase#readme "Minimalist alternative for Base. Build complex APIs with small units called plugins. Works well with most of the already existing [base][] plugins.")
- [mukla](https://www.npmjs.com/package/mukla): Small, parallel and fast test framework with suppport for async/await, promises, callbacks, streams and observables. Targets and works… [more](https://github.com/tunnckocore/mukla#readme) | [homepage](https://github.com/tunnckocore/mukla#readme "Small, parallel and fast test framework with suppport for async/await, promises, callbacks, streams and observables. Targets and works at node.js v0.10 and above.")
- [try-catch-callback](https://www.npmjs.com/package/try-catch-callback): try/catch block with a callback, used in [try-catch-core][]. Use it when you don't care about asyncness so much… [more](https://github.com/hybridables/try-catch-callback#readme) | [homepage](https://github.com/hybridables/try-catch-callback#readme "try/catch block with a callback, used in [try-catch-core][]. Use it when you don't care about asyncness so much and don't want guarantees. If you care use [try-catch-core][].")
- [try-catch-core](https://www.npmjs.com/package/try-catch-core): Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful… [more](https://github.com/hybridables/try-catch-core#readme) | [homepage](https://github.com/hybridables/try-catch-core#readme "Low-level package to handle completion and errors of sync or asynchronous functions, using [once][] and [dezalgo][] libs. Useful for and used in higher-level libs such as [always-done][] to handle completion of anything.")

## Contributing
Pull requests and stars are always welcome. For bugs and feature requests, [please create an issue](https://github.com/tunnckoCore/each-promise/issues/new).  
Please read the [contributing guidelines](CONTRIBUTING.md) for advice on opening issues, pull requests, and coding standards.  
If you need some help and can spent some cash, feel free to [contact me at CodeMentor.io](https://www.codementor.io/tunnckocore?utm_source=github&utm_medium=button&utm_term=tunnckocore&utm_campaign=github) too.

**In short:** If you want to contribute to that project, please follow these things

1. Please DO NOT edit [README.md](README.md), [CHANGELOG.md](CHANGELOG.md) and [.verb.md](.verb.md) files. See ["Building docs"](#building-docs) section.
2. Ensure anything is okey by installing the dependencies and run the tests. See ["Running tests"](#running-tests) section.
3. Always use `npm run commit` to commit changes instead of `git commit`, because it is interactive and user-friendly. It uses [commitizen][] behind the scenes, which follows Conventional Changelog idealogy.
4. Do NOT bump the version in package.json. For that we use `npm run release`, which is [standard-version][] and follows Conventional Changelog idealogy.

Thanks a lot! :)

## Building docs
Documentation and that readme is generated using [verb-generate-readme][], which is a [verb][] generator, so you need to install both of them and then run `verb` command like that

```
$ npm install verbose/verb#dev verb-generate-readme --global && verb
```

_Please don't edit the README directly. Any changes to the readme must be made in [.verb.md](.verb.md)._

## Running tests
Clone repository and run the following in that cloned directory

```
$ npm install && npm test
```

## Author
**Charlike Mike Reagent**

+ [github/tunnckoCore](https://github.com/tunnckoCore)
+ [twitter/tunnckoCore](https://twitter.com/tunnckoCore)
+ [codementor/tunnckoCore](https://codementor.io/tunnckoCore)

## Logo
The logo is [Cyclone Emoji](https://cdn.jsdelivr.net/emojione/assets/svg/1f300.svg) from [EmojiOne.com](http://emojione.com/). Released under the [CC BY 4.0](http://emojione.com/licensing/) license.

## License
Copyright © 2016-2017, [Charlike Mike Reagent](http://www.tunnckocore.tk). Released under the [MIT License](LICENSE).

***

_This file was generated by [verb-generate-readme](https://github.com/verbose/verb-generate-readme), v0.4.3, on March 19, 2017._  
_Project scaffolded using [charlike][] cli._

[always-done]: https://github.com/hybridables/always-done
[async-done]: https://github.com/gulpjs/async-done
[base]: https://github.com/node-base/base
[charlike]: https://github.com/tunnckocore/charlike
[commitizen]: https://github.com/commitizen/cz-cli
[dezalgo]: https://github.com/npm/dezalgo
[minibase]: https://github.com/node-minibase/minibase
[once]: https://github.com/isaacs/once
[p-each-series]: https://github.com/sindresorhus/p-each-series
[p-map-series]: https://github.com/sindresorhus/p-map-series
[p-map]: https://github.com/sindresorhus/p-map
[p-reduce]: https://github.com/sindresorhus/p-reduce
[p-settle]: https://github.com/sindresorhus/p-settle
[standard-version]: https://github.com/conventional-changelog/standard-version
[try-catch-core]: https://github.com/hybridables/try-catch-core
[verb-generate-readme]: https://github.com/verbose/verb-generate-readme
[verb]: https://github.com/verbose/verb

[downloads-url]: https://www.npmjs.com/package/each-promise
[downloads-img]: https://img.shields.io/npm/dt/each-promise.svg

[codeclimate-url]: https://codeclimate.com/github/tunnckoCore/each-promise
[codeclimate-img]: https://img.shields.io/codeclimate/github/tunnckoCore/each-promise.svg

[travis-url]: https://travis-ci.org/tunnckoCore/each-promise
[travis-img]: https://img.shields.io/travis/tunnckoCore/each-promise/master.svg?label=linux

[appveyor-url]: https://ci.appveyor.com/project/tunnckoCore/each-promise
[appveyor-img]: https://img.shields.io/appveyor/ci/tunnckoCore/each-promise/master.svg?label=windows

[coverage-url]: https://codecov.io/gh/tunnckoCore/each-promise
[coverage-img]: https://img.shields.io/codecov/c/github/tunnckoCore/each-promise/master.svg

[david-url]: https://david-dm.org/tunnckoCore/each-promise
[david-img]: https://img.shields.io/david/tunnckoCore/each-promise.svg

[standard-url]: https://github.com/feross/standard
[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.5"></a>
## [1.0.5](https://github.com/tunnckoCore/each-promise/compare/v1.0.4...v1.0.5) (2017-03-19)


### Bug Fixes

* **deps:** force update deps, swtich to use "native-or-another" ([4136190](https://github.com/tunnckoCore/each-promise/commit/4136190))
* **docs:** update docs ([f3696a6](https://github.com/tunnckoCore/each-promise/commit/f3696a6))
* **update:** cleanup ([9a3edb0](https://github.com/tunnckoCore/each-promise/commit/9a3edb0))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/tunnckocore/each-promise/compare/v1.0.3...v1.0.4) (2017-03-15)


### Bug Fixes

* **ci:** disallow failures on node 0.10 and node 0.12 ([29f55fb](https://github.com/tunnckocore/each-promise/commit/29f55fb)), closes [#24](https://github.com/tunnckocore/each-promise/issues/24)
* **completion:** use `redolent` for executing the functions ([60e3b84](https://github.com/tunnckocore/each-promise/commit/60e3b84))
* **handling:** always completion; handle cases when fn has callback argument ([b2323a7](https://github.com/tunnckocore/each-promise/commit/b2323a7)), closes [#23](https://github.com/tunnckocore/each-promise/issues/23)
* **hooks:** ensure finish hook is called when settle:false ([1de4ccc](https://github.com/tunnckocore/each-promise/commit/1de4ccc))
* **index.js:** pass options to redolent from promiseEach factory ([9ee5d1a](https://github.com/tunnckocore/each-promise/commit/9ee5d1a))
* **nyc:** force nyc to 100% coverage ([98ef4a3](https://github.com/tunnckocore/each-promise/commit/98ef4a3))
* **package.json:** fix lint script to lint exact files ([182865c](https://github.com/tunnckocore/each-promise/commit/182865c))
* **standard:** bump standard to v9 ([ff11c4e](https://github.com/tunnckocore/each-promise/commit/ff11c4e))
* **tests:** improve and organize tests - should work on node 0.10\n\nyou must provide a opts.Promise if you are in env that don\'t have native Promise support. If you\ndon\'t provide opts.Promise AND no native Promise support AND invalid iterable is passed to some of\nthe methods then THEY WILL THROW a TypeError with message: no native Promise and no opts.Promise.\n\nNO break-ing! ALL methods returns a promise except ONE VERY specific case when no native promise\nsupport AND invalid iterable ([f810fe4](https://github.com/tunnckocore/each-promise/commit/f810fe4))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/tunnckocore/each-promise/compare/v1.0.2...v1.0.3) (2017-02-28)


### Bug Fixes

* **ci:** update travis, add appveyor ([392ba5c](https://github.com/tunnckocore/each-promise/commit/392ba5c))
* **feature:** use try-catch-core when item is function ([59796cf](https://github.com/tunnckocore/each-promise/commit/59796cf)), closes [#9](https://github.com/tunnckocore/each-promise/issues/9)
* **options:** ensure variadic arguments ([63d98d2](https://github.com/tunnckocore/each-promise/commit/63d98d2)), closes [#19](https://github.com/tunnckocore/each-promise/issues/19)
* **package:** update npm scripts & nyc ([8e827ab](https://github.com/tunnckocore/each-promise/commit/8e827ab))
* **tests:** ensure synchronous functions throw when in "settle: true" mode ([8069464](https://github.com/tunnckocore/each-promise/commit/8069464)), closes [#3](https://github.com/tunnckocore/each-promise/issues/3)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/tunnckocore/each-promise/compare/v1.0.1...v1.0.2) (2016-11-14)


### Bug Fixes

* **package:** update deps, use "native-promise" to detect Promise ([4146b05](https://github.com/tunnckocore/each-promise/commit/4146b05))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/tunnckocore/each-promise/compare/v1.0.0...v1.0.1) (2016-11-13)


### Bug Fixes

* **readme:** update background section ([4de3ec8](https://github.com/tunnckocore/each-promise/commit/4de3ec8))





## 1.0.0 - 2016-11-13
- First release
- semantic versioning
- add docs
- add keywords
- implement

## 0.0.0 - 2016-11-10
- Initial commit

test.js
/*!
 * each-promise <https://github.com/tunnckoCore/each-promise>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

require('./test/serial')
require('./test/parallel')


test/serial.js
/*!
 * each-promise <https://github.com/tunnckoCore/each-promise>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

/* jshint asi:true */

'use strict'

var test = require('mukla')
var eachPromise = require('../index')
var Bluebird = require('bluebird')
var semver = require('semver')
var extend = require('extend-shallow')

var delay = function (fn, ms) {
  return new Bluebird(function (resolve) {
    setTimeout(resolve, ms)
  })
}

var fixtureOne = [
  123,
  function a () {
    return 456
  },
  false,
  Bluebird.resolve(555),
  'foo',
  function b () {
    return Bluebird.resolve(666)
  },
  789
]

var fixtureTwo = function () {
  return [
    delay(900).then(function () { return 1 }),
    function () { return delay(770).then(function () { throw new Error('foo') }) },
    function () { return delay(620).then(function () { return 3 }) },
    delay(800).then(function () { return 4 }),
    function () { return delay(700).then(function () { return 5 }) }
  ]
}

function factory (fnSerial) {
  test('should `.serial` resolve any type of values from iterable', function (done) {
    fnSerial(fixtureOne).then(function (res) {
      test.strictEqual(res.length, 7)
      test.deepEqual(res, [
        123,
        456,
        false,
        555,
        'foo',
        666,
        789
      ])
      done()
    }, done).catch(done)
  })

  test('should `.serial` not stop after first error if settle:true (default)', function (done) {
    fnSerial(fixtureTwo()).then(function (res) {
      test.strictEqual(res.length >= 5, true)
      test.strictEqual(res[0], 1)
      test.strictEqual(res[1].name, 'Error')
      test.strictEqual(res[2], 3)
      test.strictEqual(res[3], 4)
      test.strictEqual(res[4], 5)
      done()
    }, done).catch(done)
  })

  test('should `.serial` stop after first error if settle:false', function (done) {
    fnSerial(fixtureTwo(), { settle: false }).catch(function (err) {
      test.strictEqual(err.name, 'Error')
      test.strictEqual(err.message, 'foo')
      done()
    }).catch(done)
  })

  test('should `.serial` hooks be called', function (done) {
    var befores = 0
    var afters = 0
    var called = 0

    fnSerial(fixtureTwo(), {
      start: function () { called++ },
      beforeEach: function () { befores++ },
      afterEach: function () { afters++ },
      finish: function () { called++ },
      settle: true
    }).then(function () {
      test.strictEqual(called, 2)
      test.strictEqual(befores, 5, 'should call beforeEach hook for each')
      test.strictEqual(afters, 5, 'should call afterEach hook for each')
      done()
    }, done).catch(done)
  })

  test('should `.serial` accept `iterable` object', function (done) {
    var fixtureObj = {
      a: 123,
      b: Bluebird.reject(new Error('qux')),
      c: function () { return 456 },
      d: function () { return Bluebird.resolve(567) },
      e: function () { return new Error('zzz1') },
      f: new Error('zzz2')
    }
    fnSerial(fixtureObj).then(function (res) {
      test.strictEqual(res.length, 6)
      test.strictEqual(res[0], 123)
      test.strictEqual(res[1].name, 'Error')
      test.strictEqual(res[2], 456)
      test.strictEqual(res[3], 567)
      test.strictEqual(res[4].message, 'zzz1')
      test.strictEqual(res[5].message, 'zzz2')
      done()
    }, done).catch(done)
  })

  test('should `.serial` catch returned rejected promise', function (done) {
    fnSerial([
      function () {
        return Bluebird.reject(new Error('foo bar qux'))
      }
    ], { settle: false }).catch(function (err) {
      test.strictEqual(err.name, 'Error')
      test.strictEqual(err.message, 'foo bar qux')
      done()
    }).catch(done)
  })

  test('should `.serial` catch thrown errors', function (done) {
    var fnThrows = function () {
      zaz // eslint-disable-line no-undef, no-unused-expressions
    }

    fnSerial([fnThrows], {
      settle: false
    }).catch(function (err) {
      test.strictEqual(/zaz is not defined/.test(err.message), true)
      test.strictEqual(err.name, 'ReferenceError')
      done()
    }).catch(done)
  })

  test('should `.serial` catch returned errors', function (done) {
    var fn = function () { return new Error('xyz') }
    var options = { settle: false }

    fnSerial([fn], options).catch(function (err) {
      test.strictEqual(err.name, 'Error')
      test.strictEqual(err.message, 'xyz')
      done()
    }).catch(done)
  })

  test('should `.serial` not have flat results when `flat: false`', function (done) {
    var fns = [
      function () { return 111 },
      function () { return 222 }
    ]
    fnSerial(fns, {
      flat: false
    }).then(function (res) {
      test.strictEqual(res.length, 2)
      test.strictEqual(res[0].value, 111)
      test.strictEqual(res[1].value, 222)
      done()
    }, done).catch(done)
  }, true)

  test('should `.serial` catch when sync throw in settle:true mode', function (done) {
    fnSerial([
      function () { return 123 },
      function () { throw new Error('sync fn throws') },
      function () { return { a: 'b' } }
    ]).then(function (res) {
      test.strictEqual(res.length, 3)
      test.strictEqual(res[0], 123)
      test.strictEqual(res[1].message, 'sync fn throws')
      test.strictEqual(res[2].a, 'b')
      done()
    }, done).catch(done)
  })

  test('should call finish hook if settle:false', function (done) {
    var called = 0
    var promise = fnSerial(fixtureTwo(), {
      settle: false,
      finish: function (err) {
        test.strictEqual(err instanceof Error, true)
        called++
      }
    })

    promise
      .catch(function (er) {
        test.strictEqual(called, 1)
        test.strictEqual(er instanceof Error, true)
        done()
      })
  })
}

if (semver.lt(process.version, '0.12.0')) {
  factory(function (val, opts) {
    return eachPromise.serial(val, extend({
      Promise: Bluebird
    }, opts))
  })

  test('should on node < 0.12 - return rejected promise if not an `iterable`, but has Promise', function (done) {
    eachPromise.serial(123, { Promise: Bluebird }).catch(function (err) {
      test.strictEqual(err.name, 'TypeError')
      test.strictEqual(err.message, 'expect `iterable` to be array, iterable or object')
      done()
    }).catch(done)
  })
} else {
  factory(eachPromise.serial)

  test('should on node >= 0.12 return rejected promise if `iterable` not valid', function (done) {
    eachPromise.serial(123).catch(function (err) {
      test.strictEqual(err.name, 'TypeError')
      test.strictEqual(err.message, 'expect `iterable` to be array, iterable or object')
      done()
    }).catch(done)
  })
}


index.js
/*!
 * each-promise <https://github.com/tunnckoCore/each-promise>
 *
 * Copyright (c) Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

'use strict'

var Promize = require('native-or-another')
var utils = require('./utils')
var eachPromise = {}

/**
 * > Iterate over `iterable` in series (serially)
 * with optional `opts` (see [options section](#options))
 * and optional `mapper` function (see [item section](#item)).
 *
 * **Example**
 *
 * ```js
 * var delay = require('delay')
 * var eachPromise = require('each-promise')
 *
 * var arr = [
 *   () => delay(500).then(() => 1),
 *   () => delay(200).then(() => { throw Error('foo') }),
 *   () => delay(10).then(() => 3),
 *   () => delay(350).then(() => 4),
 *   () => delay(150).then(() => 5)
 * ]
 *
 * eachPromise
 *   .serial(arr)
 *   .then((res) => {
 *     console.log(res) // [1, Error: foo, 3, 4, 5]
 *   })
 *
 * // see what happens when parallel
 * eachPromise
 *   .parallel(arr)
 *   .then((res) => {
 *     console.log(res) // => [3, 5, Error: foo, 4, 1]
 *   })
 *
 * // pass `settle: false` if you want
 * // to stop after first error
 * eachPromise
 *   .serial(arr, { settle: false })
 *   .catch((err) => console.log(err)) // => Error: foo
 * ```
 *
 * @name   .serial
 * @param  {Array} `<iterable>` iterable object like array with any type of values
 * @param  {Function} `[mapper]` function to apply to each item in `iterable`, see [item section](#item)
 * @param  {Object} `[opts]` see [options section](#options)
 * @return {Promise} Always resolved or rejected promise
 * @api public
 */

eachPromise.serial = function eachSerial (iterable, mapper, opts) {
  var options = utils.defaults(mapper, opts)
  options = utils.extend(options, {
    serial: true
  })
  return eachPromise.each(iterable, options.mapper, options)
}

/**
 * > Iterate concurrently over `iterable` in parallel (support limiting with `opts.concurrency`)
 * with optional `opts` (see [options section](#options))
 * and optional `mapper` function (see [item section](#item)).
 *
 * **Example**
 *
 * ```js
 * var eachPromise = require('each-promise')
 *
 * var arr = [
 *   function one () {
 *     return delay(200).then(() => {
 *       return 123
 *     })
 *   },
 *   Promise.resolve('foobar'),
 *   function two () {
 *     return delay(1500).then(() => {
 *       return 345
 *     })
 *   },
 *   delay(10).then(() => 'zero'),
 *   function three () {
 *     return delay(400).then(() => {
 *       coffffnsole.log(3) // eslint-disable-line no-undef
 *       return 567
 *     })
 *   },
 *   'abc',
 *   function four () {
 *     return delay(250).then(() => {
 *       return 789
 *     })
 *   },
 *   function five () {
 *     return delay(100).then(() => {
 *       sasasa // eslint-disable-line no-undef
 *       return 444
 *     })
 *   },
 *   function six () {
 *     return delay(80).then(() => {
 *       return 'last'
 *     })
 *   }
 * ]
 *
 * // does not stop after first error
 * // pass `settle: false` if you want
 * eachPromise
 *   .parallel(arr)
 *   .then((res) => {
 *     console.log(res)
 *     // => [
 *     //   'foobar',
 *     //   'abc',
 *     //   'zero',
 *     //   'last',
 *     //   ReferenceError: sasasa is not defined,
 *     //   123,
 *     //   789,
 *     //   ReferenceError: coffffnsole is not defined
 *     //   345
 *     // ]
 *   })
 * ```
 *
 * @name   .parallel
 * @param  {Array} `<iterable>` iterable object like array with any type of values
 * @param  {Function} `[mapper]` function to apply to each item in `iterable`, see [item section](#item)
 * @param  {Object} `[opts]` see [options section](#options)
 * @return {Promise} Always resolved or rejected promise
 * @api public
 */

eachPromise.parallel = function eachParallel (iterable, mapper, opts) {
  var options = utils.defaults(mapper, opts)
  return eachPromise.each(iterable, options.mapper, utils.extend(options, {
    serial: false
  }))
}

/**
 * > Iterate over `iterable` in series or parallel (default), depending on
 * default `opts`. Pass `opts.serial: true` if you
 * want to iterate in series, pass `opts.serial: false` or does not
 * pass anything for parallel.
 *
 * **Example**
 *
 * ```js
 * var delay = require('delay')
 * var eachPromise = require('each-promise')
 *
 * var arr = [
 *   123,
 *   function () {
 *     return delay(500).then(() => 456)
 *   },
 *   Promise.resolve(678),
 *   function () {
 *     return 999
 *   },
 *   function () {
 *     return delay(200).then(() => 'foo')
 *   }
 * ]
 *
 * eachPromise
 *   .each(arr)
 *   .then(function (res) {
 *     console.log('done', res) // => [123, 678, 999, 'foo', 456]
 *   })
 * ```
 *
 * @name   .each
 * @param  {Array} `<iterable>` iterable object like array with any type of values
 * @param  {Function} `[mapper]` function to apply to each item in `iterable`, see [item section](#item)
 * @param  {Object} `[opts]` see [options section](#options)
 * @return {Promise} Always resolved or rejected promise
 * @api public
 */

eachPromise.each = function each (iterable, mapper, opts) {
  if (typeof iterable !== 'object') {
    var err = new TypeError('expect `iterable` to be array, iterable or object')
    return Promize.reject(err)
  }
  var options = utils.defaults(mapper, opts)
  return promiseEach(iterable, options)
}

/**
 * > Base iterate logic
 *
 * @param  {Array|Object} `<iterable>`
 * @param  {Object} `[opts]`
 * @return {Promise}
 * @api private
 */

function promiseEach (iterable, opts) {
  return new opts.Promise(function (resolve, reject) {
    var results = []
    var arr = Array.isArray(iterable) ? iterable : []
    arr.doneCount = 0

    if (!arr.length && typeof iterable === 'object') {
      for (var key in iterable) {
        arr.push(iterable[key])
      }
    }

    opts.concurrency = opts.serial === false ? opts.concurrency : 1
    opts.concurrency = opts.concurrency || arr.length

    opts.start()
    if (!opts.serial) {
      for (var index = 0; index < opts.concurrency; index++) {
        utils.iterator(arr, results)(opts, resolve, reject)(index)
      }
      return
    }
    utils.iterator(arr, results)(opts, resolve, reject)(0)
  })
}

module.exports = eachPromise
